alibi_detect/utils/loading.py

```diff
@@ -0,0 +1,625 @@
+# TODO - Need to modularise torch and tensorflow imports and use. e.g. has_tensorflow and has_pytorch etc
+from alibi_detect.utils.registry import registry
+from alibi_detect.utils.schemas import SupportedModels  # type: ignore[attr-defined]
+from alibi_detect.utils.validate import validate_config
+from alibi_detect.utils.tensorflow._loading import load_model as load_model_tf, \
+    prep_model_and_emb as prep_model_and_emb_tf, load_kernel_config as load_kernel_config_tf, \
+    load_optimizer as load_optimizer_tf, load_embedding as load_embedding_tf, load_detector_legacy, Detectors
+#  from alibi_detect.utils.pytorch._loading import set_device
+import tensorflow as tf  # TODO - this is currently only required for FIELDS_TO_DTYPE conversion. Remove in future.
+from transformers import AutoTokenizer
+import numpy as np
+from typing import Union, Optional, Callable, Any
+from copy import deepcopy
+from functools import partial
+from pathlib import Path
+import dill
+import os
+import toml
+from importlib import import_module
+import logging
+
+logger = logging.getLogger(__name__)
+
+REQUIRES_BACKEND = [
+    'ClassifierDrift',
+    'ClassifierUncertaintyDrift',
+    'LearnedKernelDrift',
+    'LSDDDrift',
+    'MMDDrift',
+    'RegressorUncertaintyDrift',
+    'SpotTheDiffDrift'
+]
+
+# Fields to resolve in resolve_cfg ("resolve" meaning either load local artefact or resolve @registry, conversion to
+# tuple, np.ndarray and np.dtype are dealt with separately).
+FIELDS_TO_RESOLVE = [
+    ['preprocess_fn'],
+    ['preprocess_fn', 'src'],
+    ['preprocess_fn', 'model'],
+    ['preprocess_fn', 'embedding'],
+    ['preprocess_fn', 'tokenizer'],
+    ['x_ref'],
+    ['model'],
+    ['optimizer'],
+    ['reg_loss_fn'],
+    ['kernel'],
+    ['dataset'],
+    ['kernel', 'src'],
+    ['kernel', 'proj'],
+    ['kernel', 'kernel_a', 'src'],
+    ['kernel', 'kernel_b', 'src'],
+    ['initial_diffs']
+]
+
+# Directories to amend before resolving config (fields to prepend config file dir to)
```

Comment:
Close. It's actually to prepend the `config.toml` filepath to the directories spec'd within that file.

For example, `config.toml` is stored in `filepath='my_detector'`, which we pass to `load_detector`. Within that `config.toml`, we have `preprocess_fn.model.src = 'model/'`. Since we call `load_detector` from somewhere outside of `filepath`, we need to prepend `filepath` to the directories spec'd inside the `config.toml` i.e. `preprocess_fn.model.src` needs to become `my_detector/model/`. 

The difference between `FIELDS_TO_RESOLVE` and `DIR_FIELDS` is that there are some fields which need resolving but are not directory fields themselves e.g. `embedding` is a `EmbeddingConfig` (we can't spec it as `embedding.dill` directly). The actual directory is `embedding.src`, hence why its in `DIR_FIELDS`. 

Some fields, such as `preprocess_fn.src` can be spec'd directly via `.dill`, or spec'd by setting `preprocess_fn` as a `PreprocessConfig`, hence why `preprocess_fn.src` is in both lists.

Reply:
p.s. I've added a similar comment to the code.